lib/spack/spack/patch.py

```diff
@@ -159,42 +167,43 @@ def apply(self, stage):
             os.path.dirname(stage.mirror_path),
             os.path.basename(self.url))
 
-        with spack.stage.Stage(fetcher, mirror_path=mirror) as patch_stage:
-            patch_stage.fetch()
-            patch_stage.check()
-            patch_stage.cache_local()
+        self.stage = spack.stage.Stage(fetcher, mirror_path=mirror)
+        self.stage.create()
+        self.stage.fetch()
+        self.stage.check()
+        self.stage.cache_local()
+
+        root = self.stage.path
+        if self.archive_sha256:
+            self.stage.expand_archive()
+            root = self.stage.source_path
 
-            root = patch_stage.path
+        files = os.listdir(root)
+        if not files:
             if self.archive_sha256:
-                patch_stage.expand_archive()
-                root = patch_stage.source_path
-
-            files = os.listdir(root)
-            if not files:
-                if self.archive_sha256:
-                    raise NoSuchPatchError(
-                        "Archive was empty: %s" % self.url)
-                else:
-                    raise NoSuchPatchError(
-                        "Patch failed to download: %s" % self.url)
-
-            # set this here so that path is accessible after
-            self.path = os.path.join(root, files.pop())
-
-            if not os.path.isfile(self.path):
                 raise NoSuchPatchError(
-                    "Archive %s contains no patch file!" % self.url)
-
-            # for a compressed archive, Need to check the patch sha256 again
-            # and the patch is in a directory, not in the same place
-            if self.archive_sha256 and spack.config.get('config:checksum'):
-                checker = Checker(self.sha256)
-                if not checker.check(self.path):
-                    raise fs.ChecksumError(
-                        "sha256 checksum failed for %s" % self.path,
-                        "Expected %s but got %s" % (self.sha256, checker.sum))
-
-            apply_patch(stage, self.path, self.level, self.working_dir)
+                    "Archive was empty: %s" % self.url)
+            else:
+                raise NoSuchPatchError(
+                    "Patch failed to download: %s" % self.url)
+
+        self.path = os.path.join(root, files.pop())
```

Comment:
Ok, answered my question above. Not sure if this was unclear only to me, but we currently rely on memoization to give us back always the same object from the `Spec.patches` property. Said otherwise if we remove memoization this assignment will be on the wrong patch object and we will not be able to patch the stage of an installation.

Related to failures I saw on #19501 when trying to remove `lang.memoize`. @tgamblin @scheibelp FYI